### DISCUSSION TITLE
Do I do something wrong?

### DISCUSSION THREAD

**Discussion by JanOlsmar** [oqtane/oqtane.framework#4665] [Q&A] [answered]
I have a testsite, lets call it oqtane.test.com release 5.2.3
I like to add a new site and use a new  seperate databas for this site.
I name it oqtanetest.com/newsite

I fill in this
![image](https://github.com/user-attachments/assets/c2142f4f-a901-45dd-8aa9-51d29140b15f)
(All fields is filled correct)
I get this error
An Error Occurred Creating Site. System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913. at Microsoft.EntityFrameworkCore.Infrastructure.Internal.ConcurrencyDetector.EnterCriticalSection() at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync() at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync[TSource](IAsyncEnumerable`1 asyncEnumerable, CancellationToken cancellationToken) at Microsoft.EntityFrameworkCore.Query.ShapedQueryCompilingExpressionVisitor.SingleOrDefaultAsync[TSource](IAsyncEnumerable`1 asyncEnumerable, CancellationToken cancellationToken) at Microsoft.AspNetCore.Identity.UserManager`1.FindByNameAsync(String userName) at Oqtane.Repository.UserRoleRepository.UpdateSecurityStamp(Int32 userId) in C:\Source\Projects\oqtane.framework\Oqtane.Server\Repository\UserRoleRepository.cs:line 168 at Oqtane.Repository.UserRoleRepository.AddUserRole(UserRole userRole) in C:\Source\Projects\oqtane.framework\Oqtane.Server\Repository\UserRoleRepository.cs:line 75 at Oqtane.Infrastructure.DatabaseManager.CreateSite(InstallConfig install) in C:\Source\Projects\oqtane.framework\Oqtane.Server\Infrastructure\DatabaseManager.cs:line 592

The new database is created but I cant access the new site. It is Blank no log.
I am running my test site on IIS.

**Comment by thabaum** [comment]
This sounds like it may be an issue worth reporting.

**Reply by thabaum** [reply]
if I have time today I will fire it up and run some tests.  It looks like you are using localDB or MS SQL Server?

**Reply by JanOlsmar** [reply]
I use a custom MS sql Server.

**Reply by vnetonline** [reply]
Which version are you running? .... I had the same problem in 5.2.2

5.2.3 installation works like a dream

**Reply by thabaum** [reply]
5.2.3 definitely had some added flavors to it to make it go down better.

**Reply by thabaum** [reply]
@vnetonline looks like 5.2.3 was reported in original post.  Did your issues in 5.2.2 relate to:

 `A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext.`

@JanOlsmar can you please submit your process for how you are making the two databases.  You are having Oqtane create the database correct?

**Reply by thabaum** [reply]
@JanOlsmar you have only 1 process for your application running in the app pool settings?

https://learn.microsoft.com/en-us/iis/configuration/system.applicationhost/applicationpools/add/cpu

**Reply by thabaum** [reply]
I ask because the worker process should be 1 for .NET apps from my understanding or it will cause issues.  I just want to rule this out.

**Reply by thabaum** [reply]
all sites should run under the one worker process if I am stating this correctly... something along the lines of this should be investigated first.

**Reply by JanOlsmar** [reply]
Process making two databases. First databas was created when I installed first time. Second database was made when I submitted the form above and got the error.

**Reply by thabaum** [reply]
@JanOlsmar thank you for this information.  To verify the app pool worker process has only 1 if I am not mistaken.  I will verify soon as I get a moment to log into dev station to see if I can reproduce this issue.

![image](https://github.com/user-attachments/assets/7be2fc1a-3bb4-4c83-9bae-1e84fc5c0da3)

I believe you can and should set this to 1.

**Reply by JanOlsmar** [reply]
Max 1 worker processes is my setting in the pool

**Reply by thabaum** [reply]
great thank you.  I will see what I can find out.  @sbwalker is in Florida, which declared a state of emergency today.  I hope the best for everyone dealing with that situation.  I will do my best to support until then.  On the flip side other developers contributing will be awake later in the evening that may also help provide a solution.

**Reply by JanOlsmar** [reply]
I'm in no rush, just doing a little testing to see how it worked. Hope for the best regarding Idalia.

**Comment by sbwalker** [comment, marked as answer]
@JanOlsmar it looks like this specific scenario (ie. creating a new site in a multi-tenant environment using an isolated database) may have been affected by the recent enhancements in 5.2.2 to support SecurityStamp:

Oqtane.Repository.UserRoleRepository.UpdateSecurityStamp(Int32 userId) in C:\Source\Projects\oqtane.framework\Oqtane.Server\Repository\UserRoleRepository.cs:line 168

This method relies on the .NET Identity library - which appears to be throwing the DbContext exception. Perhaps the .NET Identity library is not using DbContextFactory (which is the recommended approach in Blazor). This will require more investigation.

**Reply by sbwalker** [reply]
@thabaum yes

**Comment by sbwalker** [comment]
@JanOlsmar I just tested this locally and I was not able to reproduce the error you encountered. One difference is that I was using LocalDB rather than a full SQL Server instance. However it would be helpful to understand if you encountered the error when running in Visual Studio - or on an IIS instance.

**Reply by JanOlsmar** [reply]
This error was on a live site on IIS.
Regarding LocalDB the site show in Site setting LocalDB as type for the site. The reality is it is running on a SQL Server.

**Reply by thabaum** [reply]
@JanOlsmar can you please create an issue, along with detailed steps on how we can reproduce given the information you have provided?

I was just about to do this but I am a bit confused myself.  I would love to reproduce today if possible.  Thank you for reporting any issues like this.  We all appreciate it.

**Reply by thabaum** [reply]
@JanOlsmar P.S. I generally only review issues more than discussions so I generally do not help with discussions as much as I am more active in issues I when I have time.  I sometimes have time to review discussions, but I focus mainly on how I can personally contribute to known issues to the Oqtane Framework.  Thanks again big time!

**Reply by JanOlsmar** [reply]
@thabaum  My original post says exactly what I was doing when I got the error. Repeating this is a waste of my time. I'd rather spend it on trying to understand the error myself. If I have new information I will post it here.

**Reply by thabaum** [reply]
@JanOlsmar gotcha, thank you

**Reply by thabaum** [reply]
@JanOlsmar I created the issue for you please review as this is where I will be following up on the issue after I can reproduce and update details so others can join in on the fun :)

https://github.com/oqtane/oqtane.framework/issues/4667

Thank you!

**Reply by thabaum** [reply]
> This error was on a live site on IIS. Regarding LocalDB the site show in Site setting LocalDB as type for the site. The reality is it is running on a SQL Server.

@JanOlsmar So the issue is with using Microsoft SQL Server and NOT LocalDB correct?

FYI Just a thumbs up is enough of a reply.  I am setting up this environment now to help report anything I might be able to add to this discussion and the issue I created for it.

**Reply by thabaum** [reply]
@JanOlsmar one other question is does this "production" environment use a control panel such as Plesk or is it a stand alone server or Azure?  This should help a lot in reproducing this issue.

**Reply by thabaum** [reply]
For example this issue here was a control panel hosting issue: https://github.com/oqtane/oqtane.framework/discussions/4646

**Reply by JanOlsmar** [reply]
As stated from beginning SQL server was used. (However Site setting tells me it is LocalDB and I have not yet checked if this value is used in the process) By the way I have never used LocalDB when running a site in IIS.
However I thought I should try this today. I tryed to make new website Oqtane an used LocalDB as database but got EXACT same error as  described in this tread. I got the error before the new site was created.
I use my own server.
LocalDB works fine in developement mode)

**Reply by thabaum** [reply]
```
  "AvailableDatabases": [
    {
      "Name": "LocalDB",
      "ControlType": "Oqtane.Installer.Controls.LocalDBConfig, Oqtane.Client",
      "DBType": "Oqtane.Database.SqlServer.SqlServerDatabase, Oqtane.Database.SqlServer"
    },
    {
      "Name": "SQL Server",
      "ControlType": "Oqtane.Installer.Controls.SqlServerConfig, Oqtane.Client",
      "DBType": "Oqtane.Database.SqlServer.SqlServerDatabase, Oqtane.Database.SqlServer"
    },
```

In the appsettings.json file it shows the same "Server DB Type" for LocalDB and MS SQL, this is an upgraded instance I have setup using Windows 11 IIS and Developer Edition of MS SQL Server to hopefully reproduce adjusting.  I will report back shortly once I can make this happen as easy as you can. I can also spin up a Windows 2019 server if needed.

This instance seems to be connected to LocalDB instance and you say you can reproduce after reviewing so I will just use this for now.

**Reply by thabaum** [reply]
I was able to reproduce the error message

**Reply by JanOlsmar** [reply]
Maybe you didnt understand. I could not make a site with LocalDB I got an error. Same error as in this tread.

**Reply by thabaum** [reply]
I have updated issue #4667 with relating screenshots verifying your conclusions

**Reply by thabaum** [reply]
@JanOlsmar I also updated the issue with how to reproduce so @sbwalker can do this from his laptop if needed.

**Reply by thabaum** [reply]
thanks again for reporting, I will dig into code as I can as he gave us some clues to work on with latest updates to the DB Context Factory, however I cannot verify this until I work deeper in the framework source and push tests to the example production site I have for testing things like this.

**Reply by thabaum** [reply]
This helps get the ball rolling a little faster, thanks again for reporting and please let us know if you find anything more out that may help.

**Reply by thabaum** [reply]
@JanOlsmar thanks again for additional support.  one last question, sorry to ask as I have not tested via MS SQL Server but did get the error in LocalDB sub-site creation.  You mentioned the database is created, is it verified a blank database in MS SQL Server Management Studio second database is created, but it is empty?  Not just included Oqtane appsettings.json file but actually a database was created?

I may have time later to review this more, but I don't want to have to do this if it is unnecessary since you say you already have been through that process trying to create a second site with a second database created using MS SQL Server (not LocalDB)

**Reply by thabaum** [reply]
@JanOlsmar disregard question as I believe once issue #4667 is resolved so with this issue, thanks for reporting!

**Comment by sbwalker** [comment]
@thabaum the issue related to Installation on IIS #4667 has been resolved... however the issue of creating an additional Site with a separate Database(Tenant) which @JanOlsmar reported in this Discussion thread is unrelated, and still appears to be an issue (ie. I can reproduce it). I cannot reproduce this issue when running in Visual Studio (on IIS Express) so it is very confusing.

```
An Error Occurred Migrating The Database For Tenant test. This Is Usually Related To Database Permissions, Connection String Mappings, Or The Database Not Being In A Supported State. System.InvalidOperationException: No database provider has been configured for this DbContext. A provider can be configured by overriding the 'DbContext.OnConfiguring' method or by using 'AddDbContext' on the application service provider. If 'AddDbContext' is used, then also ensure that your DbContext type accepts a DbContextOptions<TContext> object in its constructor and passes it to the base constructor for DbContext.
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.Initialize(IServiceProvider scopedProvider, DbContextOptions contextOptions, DbContext context)
   at Microsoft.EntityFrameworkCore.DbContext.get_ContextServices()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.GetRelationalService[TService](IInfrastructure`1 databaseFacade)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.Migrate(DatabaseFacade databaseFacade)
   at Oqtane.Infrastructure.DatabaseManager.MigrateTenants(InstallConfig install) in C:\Source\Projects\oqtane.framework\Oqtane.Server\Infrastructure\DatabaseManager.cs:line 377 
```

**Reply by thabaum** [reply]
This is different error than originally reported.

It looks like a permission issue, which could be something different entirely due to the limits the app service may have on creating a database in the app data folder it gets created in.

**Reply by thabaum** [reply]
The best way to test would be MS SQL with a user that has full access to create a database

**Reply by thabaum** [reply]
'An Error Occurred Migrating The Database For Tenant test. This Is Usually Related To Database Permissions, Connection String Mappings, Or The Database Not Being In A Supported State. System.InvalidOperationException: No database provider has been configured for this DbContext. A provider can be configured by overriding the 'DbContext.OnConfiguring' method or by using 'AddDbContext' on the application service provider. If 'AddDbContext' is used, then also ensure that your DbContext type accepts a DbContextOptions<TContext> object in its constructor and passes it to the base constructor for DbContext.'

This is interesting, I would still test against MS SQL or give the folder that creates database for LocalDB full permissions from the app pool user identity

**Reply by thabaum** [reply]
this way you can rule out "permissions" as an issue.
edited:
I always wondered how LocalDB permissions would work in this experience FYI this is something worthy of reviewing for production IIS/Apache database support.  I will try to add our findings to the Oqtane Documentation along with a few more issues we are now aware of with WebDAV enabled in control panels for example if no one beats me to it... ;)

**Reply by thabaum** [reply]
https://learn.microsoft.com/en-us/sql/database-engine/configure-windows/sql-server-express-localdb?view=sql-server-ver16

> The LocalDB setup program uses the SqlLocalDB.msi program to install the necessary files on the computer. Once installed, LocalDB is an instance of SQL Server Express that can create and open SQL Server databases. The system database files for the database are stored in the local AppData path, which is normally hidden. For example, C:\Users\<user>\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\LocalDBApp1\. User database files are stored where the user designates, typically somewhere in the C:\Users\<user>\Documents\ folder.

**Reply by sbwalker** [reply]
@thabaum #4676 fixes this issue (logged as #4675). Basically the problem was that when creating a new database it dynamically adds a new connection string to appsettings.json - however the call to the AddOrUpdateSetting() method was not setting the "reload" parameter to true. As a result, when it tried to create the DbContext it could not retrieve the connection string (because it was using an older cached version of IConfiguration). What I don't understand is why this did not throw an exception in Visual Studio. Perhaps when debugging it automatically reloads appsettings.json  changes.

**Reply by thabaum** [reply]
@sbwalker gotcha, glad you found the issue.  This makes sense and I believe VS probably is working as you described which maybe related to hot reload?  Maybe running without hot reload would create this error?

**Reply by thabaum** [reply]
The permission issue if using integrated security permissions and should have all permissions needed.  So that would rule that out.

**Comment by JanOlsmar** [comment]
I have done a couple of test of my original error  IIS and  Sql Server. I never use LocalDB on IIS.
I have tested folder and subsite. Same result and error.
The databases that are created is populated and fine. I can change connectionstrings and get a site running. 
the appsetting file is updated with new connection string.
But when browsing the new site eg sub.domain.com or www.domain.com/folder a  blank page is displayed
<html>
<head>
<body>
No error in any logs (IIS and Site) whe browsing the new site.

**Comment by sbwalker** [comment]
@JanOlsmar are you saying that you got the latest code from the Dev branch (with the latest fix), compiled it, deployed it to an IIS instance, and you are still experiencing the problem?

**Reply by JanOlsmar** [reply]
No was not aware there was a fix for this issue. Still on  5.2.3.
Will try test latest code.

**Reply by sbwalker** [reply]
I posted this above but it probably got lost in all the other posts "https://github.com/oqtane/oqtane.framework/pull/4676 fixes this issue (logged as https://github.com/oqtane/oqtane.framework/issues/4675). Basically the problem was that when creating a new database it dynamically adds a new connection string to appsettings.json - however the call to the AddOrUpdateSetting() method was not setting the "reload" parameter to true. As a result, when it tried to create the DbContext it could not retrieve the connection string (because it was using an older cached version of IConfiguration). What I don't understand is why this did not throw an exception in Visual Studio. Perhaps when debugging it automatically reloads appsettings.json changes."

**Reply by JanOlsmar** [reply]
No problem :) Works fine with Dev branch.

**Reply by sbwalker** [reply]
I am happy this problem is resolved. Thank you for reporting it!

**Comment by mdmontesinos** [comment]
FYI, the same also happens when trying to deploy in Azure (Web App + SQL Server)

![image](https://github.com/user-attachments/assets/9a9e3a6b-a548-483f-ad7a-99242ff6488f)

This is with the Install zip from version 5.2.3, so it's not possible to deploy that version in Azure. An urgent 5.2.4 release (that contains the mentioned fix) is required.
